### PR TITLE
GpxMultiSegmentsApproximation improvements

### PR DESCRIPTION
**Pull request by RZR-UA**
1) fixApproximateTestFree without breakSegment (gpx loops, etc)
2) RouteSegmentAppr.onewayPenalty (bypass oneway=yes)

All 4 new tests passed now.

### CODE REVIEW

**Comment by RZR-UA** [comment]
https://github.com/osmandapp/OsmAnd/issues/20280

File: OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java [hunk 1]
```diff
@@ -119,15 +123,24 @@ private void addSegmentInternal(RouteSegmentAppr last, RouteSegment sg, List<Rou
...
 	private void addSegment(RouteSegmentAppr last, RouteSegment sg, List<RouteSegmentAppr> connected) {
 		if (sg == null) {
 			return;
 		}
-		if (sg.getRoad().getId() != last.segment.road.getId() || 
-				Math.min(sg.getSegmentStart(), sg.getSegmentEnd()) != 
-				Math.min(last.segment.getSegmentStart(), last.segment.getSegmentEnd())) {
+
+		int prev = last.gpxStart;
+		int next = Math.min(last.gpxNext(), gpxPoints.size() - 1);
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java:133]: too complicated , definitely don't like this loops vs simple existing code doesn't make sense

File: OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java [hunk 2]
```diff
@@ -21,60 +21,66 @@ public class GpxMultiSegmentsApproximation {
...
+			this.onewayPenalty = calcOnewayPenalty(minPointApproximation);
 		}
-		
-		private RouteSegmentAppr(RouteSegmentAppr parent, RouteSegment segment) {
+
+		private RouteSegmentAppr(RouteSegmentAppr parent, RouteSegment segment, float minPointApproximation) {
 			this.parent = parent;
 			this.segment = segment;
 			this.gpxStart = parent.gpxStart + parent.gpxLen;
+			this.onewayPenalty = calcOnewayPenalty(minPointApproximation);
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java:71]: wrong place to calculate oneway

File: OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java [hunk 3]
```diff
@@ -119,15 +123,24 @@ private void addSegmentInternal(RouteSegmentAppr last, RouteSegment sg, List<Rou
...
-	
+
 	private void addSegment(RouteSegmentAppr last, RouteSegment sg, List<RouteSegmentAppr> connected) {
 		if (sg == null) {
 			return;
 		}
-		if (sg.getRoad().getId() != last.segment.road.getId() || 
-				Math.min(sg.getSegmentStart(), sg.getSegmentEnd()) != 
-				Math.min(last.segment.getSegmentStart(), last.segment.getSegmentEnd())) {
+
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java:131]: Correct code for oneway is
```
int oneway = gctx.ctx.getRouter().isOneWay(sg.getRoad());
		if ((sg.isPositive() && oneway < 0) || (!sg.isPositive() && oneway > 0)) {
			// don't allow passing wrong way
			return;
		}
```

File: OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java (hunk 2 above)
Comment (RZR-UA) [reply OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java:71]: agreed. moved to better place

File: OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java (hunk 1 above)
Comment (RZR-UA) [reply OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java:123]: the `loops` was your code initially:

```
		if (sg.getRoad().getId() != last.segment.road.getId() ||
				sg.getSegmentStart() != last.segment.getSegmentStart() // 1. allow loops for rec tracks
//				Math.min(sg.getSegmentStart(), sg.getSegmentEnd()) != 
//				Math.min(last.segment.getSegmentStart(), last.segment.getSegmentEnd()) // 2. don't allow loops
				) {
```

Comment (RZR-UA) [reply OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java]: created method to keep addSegment simple

**Pull request closed without merging**